nucypher/blockchain/eth/sol/source/contracts/UserEscrow.sol

```diff
@@ -13,22 +13,29 @@ import "contracts/NuCypherToken.sol";
 contract UserEscrowLibraryLinker is Ownable {
 
     address public target;
+    bytes32 public secretHash;
 
     /**
     * @param _target Address of the library contract
+    * @param _newSecretHash Secret hash (keccak256)
     **/
-    constructor(address _target) public {
+    constructor(address _target, bytes32 _newSecretHash) public {
         require(_target != 0x0);
         target = _target;
+        secretHash = _newSecretHash;
     }
 
     /**
     * @notice Upgrade library
     * @param _target New contract address
+    * @param _secret Secret for proof of contract owning
+    * @param _newSecretHash New secret hash (keccak256)
     **/
-    function upgrade(address _target) public onlyOwner {
+    function upgrade(address _target, bytes _secret, bytes32 _newSecretHash) public onlyOwner {
```

Comment:
Curious - why is the secret `bytes`?  So it can be an arbitrary length?  I noticed there is a constant `SECRET_LENGTH`...

Reply:
SECRET_LENGTH is just for tests. We can choose any length and change it every time